File: backend/app/schemas/weather.py
from typing import Optional

from pydantic import BaseModel


class WeatherSummaryResponse(BaseModel):
    temperature: float
    feels_like: float
    humidity: int
    pressure: int
    condition_description: str
    condition_icon: str
    wind_speed: float
    clouds: int
    aqi: Optional[int] = None
    aqi_label: Optional[str] = None
    kp_index: Optional[float] = None
    storm_severity: Optional[str] = None
    city: str


class UVForecastPoint(BaseModel):
    value: float
    forecast_at: str


class UVData(BaseModel):
    value: float
    risk_level: str
    risk_label: str
    forecast: list[UVForecastPoint] = []


class CircadianData(BaseModel):
    sunrise: Optional[str] = None
    sunset: Optional[str] = None
    day_length_hours: float
    cloud_cover_percent: int
    effective_light_hours: float
    quality: str
    quality_label: str


class AQForecastPoint(BaseModel):
    aqi: int
    label: str
    measured_at: str


class AirQualityData(BaseModel):
    aqi: int
    label: str
    co: float
    no: float
    no2: float
    o3: float
    so2: float
    pm2_5: float
    pm10: float
    nh3: float
    forecast: list[AQForecastPoint] = []


class StormForecastPoint(BaseModel):
    predicted_kp: float
    g_scale: str
    severity: str
    is_storm: bool
    period_start: str
    period_end: str


class MagneticStormData(BaseModel):
    kp_index: float
    g_scale: str
    severity: str
    is_storm: bool
    health_impact_level: Optional[str] = None
    affected_systems: list[str] = []
    recommendations: list[str] = []
    disclaimer: Optional[str] = None
    forecast: list[StormForecastPoint] = []


class WeatherDetailResponse(BaseModel):
    weather: Optional[WeatherSummaryResponse] = None
    uv: Optional[UVData] = None
    circadian: Optional[CircadianData] = None
    air_quality: Optional[AirQualityData] = None
    magnetic_storm: Optional[MagneticStormData] = None
    city: str

File: frontend/src/types/weather.ts
export interface WeatherSummary {
  temperature: number
  feels_like: number
  humidity: number
  pressure: number
  condition_description: string
  condition_icon: string
  wind_speed: number
  clouds: number
  aqi: number | null
  aqi_label: string | null
  kp_index: number | null
  storm_severity: string | null
  city: string
}

export interface UVForecastPoint {
  value: number
  forecast_at: string
}

export interface UVData {
  value: number
  risk_level: string
  risk_label: string
  forecast: UVForecastPoint[]
}

export interface CircadianData {
  sunrise: string | null
  sunset: string | null
  day_length_hours: number
  cloud_cover_percent: number
  effective_light_hours: number
  quality: string
  quality_label: string
}

export interface AQForecastPoint {
  aqi: number
  label: string
  measured_at: string
}

export interface AirQualityData {
  aqi: number
  label: string
  co: number
  no: number
  no2: number
  o3: number
  so2: number
  pm2_5: number
  pm10: number
  nh3: number
  forecast: AQForecastPoint[]
}

export interface StormForecastPoint {
  predicted_kp: number
  g_scale: string
  severity: string
  is_storm: boolean
  period_start: string
  period_end: string
}

export interface MagneticStormData {
  kp_index: number
  g_scale: string
  severity: string
  is_storm: boolean
  health_impact_level: string | null
  affected_systems: string[]
  recommendations: string[]
  disclaimer: string | null
  forecast: StormForecastPoint[]
}

export interface WeatherDetail {
  weather: WeatherSummary | null
  uv: UVData | null
  circadian: CircadianData | null
  air_quality: AirQualityData | null
  magnetic_storm: MagneticStormData | null
  city: string
}
